pkg/controllers/workapplier/controller.go

```diff
@@ -487,27 +472,122 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
 	}
-	err := r.spokeClient.Delete(ctx, &appliedWork, &client.DeleteOptions{PropagationPolicy: &deletePolicy})
-	switch {
-	case apierrors.IsNotFound(err):
-		klog.V(2).InfoS("The appliedWork is already deleted", "appliedWork", work.Name)
-	case err != nil:
-		klog.ErrorS(err, "Failed to delete the appliedWork", "appliedWork", work.Name)
-		return ctrl.Result{}, controller.NewAPIServerError(false, err)
-	default:
-		klog.InfoS("Successfully deleted the appliedWork", "appliedWork", work.Name)
+	// Get the AppliedWork object
+	if err := r.spokeClient.Get(ctx, types.NamespacedName{Name: work.Name}, &appliedWork); err != nil {
+		if apierrors.IsNotFound(err) {
+			klog.V(2).InfoS("The appliedWork is already deleted, removing the finalizer from the work", "appliedWork", work.Name)
+			return r.removeWorkFinalizer(ctx, work)
+		}
+		klog.ErrorS(err, "Failed to get AppliedWork", "appliedWork", work.Name)
+		return ctrl.Result{Requeue: true}, controller.NewAPIServerError(false, err)
 	}
-	controllerutil.RemoveFinalizer(work, fleetv1beta1.WorkFinalizer)
 
+	// Delete the appliedWork object.
+	if appliedWork.DeletionTimestamp.IsZero() {
+		// Update the owner reference in the AppliedWork object.
+		if err := r.updateOwnerReference(ctx, &appliedWork); err != nil {
+			klog.ErrorS(err, "Failed to update owner reference in the appliedWork", "appliedWork", work.Name)
+			return ctrl.Result{}, controller.NewAPIServerError(false, err)
+		}
+		klog.V(2).InfoS("Deleting the appliedWork", "appliedWork", work.Name)
+		if err := r.spokeClient.Delete(ctx, &appliedWork, &client.DeleteOptions{PropagationPolicy: &deletePolicy}); err != nil {
+			if apierrors.IsNotFound(err) {
+				klog.V(2).InfoS("The appliedWork is already deleted", "appliedWork", work.Name)
+				return r.removeWorkFinalizer(ctx, work)
+			}
+			klog.V(2).ErrorS(err, "Failed to delete the appliedWork", "appliedWork", work.Name)
+			return ctrl.Result{}, controller.NewAPIServerError(false, err)
+		}
+	}
+	klog.V(2).InfoS("AppliedWork deletion in progress", "appliedWork", work.Name)
+	return ctrl.Result{Requeue: true}, fmt.Errorf("AppliedWork %s is being deleted, waiting for the deletion to complete", work.Name)
+}
+
+// updateOwnerReference updates the AppliedWork owner reference in the manifest objects.
+// It changes the `blockOwnerDeletion` field to true, so that the AppliedWork can be deleted in cases where
+// the other owner references do not exist or are invalid.
+// https://kubernetes.io/docs/concepts/overview/working-with-objects/owners-dependents/#owner-references-in-object-specifications
+func (r *Reconciler) updateOwnerReference(ctx context.Context, appliedWork *fleetv1beta1.AppliedWork) error {
+	appliedWorkOwnerRef := &metav1.OwnerReference{
+		APIVersion: fleetv1beta1.GroupVersion.String(),
+		Kind:       "AppliedWork",
+		Name:       appliedWork.Name,
+		UID:        appliedWork.UID,
+	}
+
+	for _, res := range appliedWork.Status.AppliedResources {
```

Comment:
Hi Britania! This part might have some potential consistency issues; the `AppliedResources` on AppliedWork objects are updated only at the end of the reconciliation loop on a best effort basis; it might not incude all the applied manifests if, say, the status update is skipped/interruptted. The work object status is, instead, designed to be always consistent.